### PR TITLE
NIAD-1991: Practitioners, PractitionerRole and Organizations from Agents

### PR DESCRIPTION
**Pull request by rorysproule1**
[NIAD-1991](https://gpitbjss.atlassian.net/browse/NIAD-1991)

The AgentDirectory Mapper takes the AgentDirectory, with it's list of Agents and generates Practitioner/Organization/PractitionerRole resources from each of these Agents as appropriate. When finished it returns the list of these new resources.

The overarching logic for which resources are generated is as follows:

1. If the Agent has an agentPerson and representedOrganization -> then Practitioner, Organization & PractitionerRole are generated
2. If the Agent only has an agentPerson -> then only a Practitioner is generated
3. If the Agent only has an agentOrganization -> then only an Organization is generated.

The Organization generated by agentOrganizations & representedOrganizations is mostly the same, expect the agentOrg creates a type atrribute and the representedOrg has an '-ORG' suffix on it's id

### CODE REVIEW

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/AgentDirectoryMapper.java [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+        practitioner.setId(id);
+        practitioner.getMeta().getProfile().add(new UriType(PRACT_META_PROFILE));
+        practitioner.setName(getPractitionerName(agentPerson.getName()));
+
+        return practitioner;
+    }
+
+    private List<HumanName> getPractitionerName(PN name) {
+        var nameList = new ArrayList();
```
Comment (bchotkowski) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/AgentDirectoryMapper.java:95]:
```suggestion
        var nameList = new ArrayList<HumanName>();
```
same here in the declaration of ArrayList?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/AgentDirectoryMapper.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    private String getTelecomValue(TEL telecom) {
+        var value = telecom.getValue();
+        if (value != null) {
+            return isWorkPlaceValue(telecom.getUse()) ? stripTelPrefix(value) : value;
+        }
+
+        return null;
+    }
+
+    private boolean isWorkPlaceValue(List<CsTelecommunicationAddressUse> use) {
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/AgentDirectoryMapper.java:175]: I think this method, `stripTelPrefix` and some address-related methods are in LocationMapper as well. Would it be worth to create a util class for those, which could hold all useful address & telecom methods?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/AgentDirectoryMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+    }
+    private void assertTelecom(ContactPoint telecom, String value) {
+        assertThat(telecom.getSystem()).isEqualTo(ContactPointSystem.PHONE);
+        assertThat(telecom.getUse()).isEqualTo(ContactPointUse.WORK);
+        assertThat(telecom.getRank()).isEqualTo(TELECOM_RANK);
+        assertThat(telecom.getValue()).isEqualTo(value);
+    }
+
+    @SneakyThrows
+    private RCMRMT030101UK04AgentDirectory unmarshallCodeElement(String fileName) {
```
Comment (mrogowska) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/AgentDirectoryMapperTest.java:318]: I think the name could be `unmarshallAgentDirectoryElement`

**Review by mrogowska — CHANGES_REQUESTED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/AgentDirectoryMapper.java (hunk 2 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/AgentDirectoryMapper.java]: Great idea! I've created 2 seperate Util classes for these 😄

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/AddressUtil.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package uk.nhs.adaptors.pss.translator.util;
+
+import org.hl7.fhir.dstu3.model.Address;
+import org.hl7.v3.AD;
+
+import io.micrometer.core.instrument.util.StringUtils;
+
+public class AddressUtil {
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/AddressUtil.java:8]: Could you please add unit tests for the new classes?

**Review by mrogowska — CHANGES_REQUESTED** [review]

**Review by mrogowska — APPROVED** [review]

**Review by bchotkowski — APPROVED** [review]

**Pull request merged**